### PR TITLE
Use shared webpack runtime

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue #2444

## Relevant technical choices

* Changes configuration of how `i18n` is built so that `require( '@wordpress/i18n' ) === window.googlesitekit.i18n` without requiring extra complications with externals and shim modules

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
@aaemnnosttv, please, revert i18n related changes because it breaks internalization and causes a new regression.

**Comment by aaemnnosttv** [comment]
@eugene-manuilov thanks for pointing out the issue with i18n (wish this was easier to test). It works now, although using a slightly different approach which is closer to what Gutenberg uses in Webpack.

image

**Review by eugene-manuilov — DISMISSED** [review]
Awesome! I like the new approach! :+1:

**Review by tofumatt — DISMISSED** [review]
more-red-than-green

This is so awesome. Great job, looks good! 👍🏻

**Review by felixarntz — CHANGES_REQUESTED** [review]
Technically looks good for the most part, however this PR breaks JS translations - I'm not sure for which reason, but when I saw it changed the generation of `googlesitekit-i18n.js` and how the imports resolve, it got me worried. I tested it, and indeed JS translations load with `develop`, but they don't with this branch.

IMO we shouldn't need to touch `googlesitekit-i18n` at all here - doesn't that go beyond the scope of work here?

File: .storybook/webpack.config.js [hunk 1]
```diff
@@ -1,22 +1,46 @@
...
 	const siteKitPackageAliases = mapValues(
-		mainConfig.siteKitExternals,
+		siteKitExternals,
 		( [ global, api ] ) => {
-			// Revert "@wordpress/i18n: [ googlesitekit, i18n ]" external back to the original @wordpress/i18n.
-			if ( global === 'googlesitekit' ) {
-				if ( api === 'i18n' ) {
-					return require.resolve( '@wordpress/i18n' );
-				}
-			}
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply .storybook/webpack.config.js:18]: It simplifies the setup needed to provide that library on our global. The old configuration didn't shim it in quite the same way as the others (sourcing exports from the global if it existed) but other than that, we still needed to create a `'@wordpress/i18n__non-shim': require.resolve( '@wordpress/i18n' ),` alias for it to work around the import.

It's not strictly needed for this issue but seemed relevant due to its "shim" alias.

**Comment by aaemnnosttv** [comment]
> however this PR breaks JS translations - I'm not sure for which reason

@felixarntz I think it may be related to the translation loader plugin – @eugene-manuilov and I both ran into issues with this initially and found that it didn't work properly in some cases. Perhaps that's why translations worked for me and not for you.  Odd that it would be much different between develop and this branch though (unless maybe you hadn't merged develop in it yet?).

At any rate, I've reverted the i18n changes and translations still work for me 😄  It does make this PR a little smaller though, and we can revisit our API library bundling in a separate issue.

**Review by felixarntz — COMMENTED** [review]
@aaemnnosttv I'm not sure what's going on here, but for me the translations still don't work while they _do_ work against latest `develop`. Let's go over this one tomorrow.

**Comment by eugene-manuilov** [comment]
> aaemnnosttv I'm not sure what's going on here, but for me the translations still don't work while they _do_ work against latest `develop`. Let's go over this one tomorrow.

@felixarntz do the following:

- Switch back to English if you are on another language;
- Delete the `wp-content/languages` folder;
- Delete your current version of the `Site Kit Language Loader` plugin;
- Install and activate this version of the `Site Kit Language Loader` plugin: 
[google-site-kit-language-loader.zip](https://github.com/google/site-kit-wp/files/5665735/google-site-kit-language-loader.zip)
- Checkout the `fix/2444-multiple-webpack-runtimes` branch;
- Build development assets;
- Go to the settings and select Espanol language;
- Go to the `Dashboard` > `Updates` page (`/wp-admin/update-core.php`) and click on the button to update translations (usually it is located at the very bottom of that page).
- Go to the site kit dashboard and enjoy :es: translations :partying_face:

**Review by felixarntz — CHANGES_REQUESTED** [review]
@aaemnnosttv @eugene-manuilov The reason my version of the language loader plugin doesn't work with this PR (and the reason Eugene's version doesn't work with `develop`) is that the generated asset file names are different in format between the two branches:
* `develop` has the hash separated by a dot at all times.
* This branch has the hash mostly separated by a dash, but sometimes with a dot too.

This is likely due to the `runtimeChunk` change: I think we'll be able to fix this by updating `output.chunkFilename` to also use a dot as separator, which I think makes sense for consistency. Continuing to use a dot (over the dash) makes sense to me because the asset handles themselves mainly include dashes and (with one exception) no dots.

So let's update `output.chunkFilename`. Then Eugene's language loader will need to be updated to split accordingly by dot instead of dash, but that should be trivial. It's generally simpler than the previous language loader plugin version, so overall I'd advise to declare Eugene's version the "canonical recommendation" 😆

**Comment by felixarntz** [comment]
@eugene-manuilov After the above changes to this branch, the relevant area of your language loader could then be updated to the following:
```php
	$hndl = explode( '.', basename( $src ) );
	array_pop( $hndl ); // Remove .js extension.
	array_pop( $hndl ); // Remove hash.
	$hndl = implode( '.', $hndl );
```

Then it will work across both branches.

**Comment by aaemnnosttv** [comment]
Nice catch @felixarntz – it seems like using the runtime chunk treats the entry points as chunks instead (perhaps that's the main difference between chunks and an entry point?). I found changing the `output.chunkFilename` to use the same format results in the same filename format as before so this should be good now 👍

**Review by felixarntz — APPROVED** [review]
Awesome! Seems to fix all the "Maximum call stack size exceeded" errors, and works now maintaining the previous file name format.

**Pull request merged**